### PR TITLE
Support for abbreviated IPv4 subnet notation

### LINKED ISSUE #65: Support for abbreviated IPv4 subnet notation?
I recently started using this library and while adding unit tests for the new functionality noticed a few corner cases.

Like with IPv6, IPv4, too, has some abbreviated subnet notations. Like with the `::` in IPv6 addresses, these are only supported at the end of the IP string and assumes that any missing bits get expanded to all zeros. This abbreviated notation is, for example, supported in both [GNU](https://linux.die.net/man/3/inet_network) and [BSD libc](https://www.freebsd.org/cgi/man.cgi?query=inet_net&sektion=3&apropos=0&manpath=FreeBSD+12.2-RELEASE+and+Ports). While this abbreviated notation is therefore widely supported in many utilities that link the libc, I couldn't find RFC documentation that specifically formalizes it, but it is used in RFCs to describe networks, for example in [RFC 1338, section 5](https://datatracker.ietf.org/doc/html/rfc1338#section-5).

Here are some example abbreviations and their equivalents in the currently supported notation:
- `'127/24'` ~ `127.0.0.*`
- `'10/8'` ~ `10.0.0.0/8`
- `'10.10.10/24'` ~ `10.10.10.0/24`

This is certainly not blocking anyone and one can easily use a different network notation to get the desired result. Many folks don't even know of these abbreviations. Given that implementing this will increase complexity of the library and therefore introduces risks, I can totally understand, if this is considered to be outside of the scope of this library, which would conclude this issue, documenting that these notations are not supported.

If you, on the other hand, would consider reviewing a merge request for this. I'd be willing to invest a few hours to raise one for such a feature + test cases.

### PR DESCRIPTION
**Pull request by elrido**
This change extends the supported formats of the range's `fromString()` with the abbreviated IPv4 CIDR subnet notation. Some examples for these and the expected lowest and highest IPs can be found in the associated test case.

closes #65

### CODE REVIEW

File: src/Range/Pattern.php [hunk 1]
```diff
@@ -77,7 +77,7 @@ public function __toString()
      */
     public static function fromString($range, $supportNonDecimalIPv4 = false)
     {
-        if (!is_string($range) || strpos($range, '*') === false) {
```
Comment (mlocati) [inline src/Range/Pattern.php:80]: Pattern ranges are in the form `1.2.3.*`, so if a string does not contain an asterisk, it's not a `Range\Pattern` range.

File: src/Range/Pattern.php [hunk 2]
```diff
@@ -118,6 +118,29 @@ public static function fromString($range, $supportNonDecimalIPv4 = false)
 
             return new static($fromAddress, $toAddress, $asterisksCount);
         }
+        if (strpos($range, '/') !== false) {
```
Comment (mlocati) [inline src/Range/Pattern.php:121]: If a string contains `/` it may be a `Range\Subnet` pattern, isn't it?

Comment (elrido) [reply src/Range/Pattern.php:121]: Did I implement this in the wrong class? Sorry, I'll move it over there, then.

Comment (elrido) [reply src/Range/Pattern.php]: Thank you for the pointer, the `Range\Subnet` needed only a tiny adjustment. Reverted the changes in `Range\Pattern`.

**Comment by mlocati** [comment]
Much cleaner, thanks!

BTW, one more question: is `127.1/32` the same as `127.0.0.1/32` or `127.1.0.0/32`?

I'm asking because for addresses we have that `127.1` is the same as `127.0.0.1`:

```sh
$ ping 127.1
PING 127.1 (127.0.0.1) 56(84) bytes of data.
```

(See also https://github.com/mlocati/ip-lib/issues/59)

**Comment by elrido** [comment]
Good question! I'd have expected it to resolve to 127.1.0.0, but the ping example clearly shows that (g)libc handles this differently.

I'll do some experimentation and try to find some references for the abbreviated syntax and adjust the MR with additional test cases and changes, if necessary, to ensure the behaviour is consistent for subnets and single IPs and stays as close as possible to that of other implementations.

**Comment by elrido** [comment]
The logic now got corrected to insert the zero bytes _before_ the last byte, if the given IPv4 address string has less then 3 dots. Since this now applies to both single IP and range parsing, it did break a number of unit tests. Please review those changes in particular

I moved some of the invalid cases to now be valid, so this has become a breaking change. If you would prefer it to be backwards compatible, I could add a `$supportAbbreviations = false` feature flag to the `\IPLib\Address\IPv4::fromString()` (and maybe also to the factories and the range classes), for it to be disabled by default.

**Comment by mlocati** [comment]
I'm not sure that subnets are abbreviated like single IPs: from https://datatracker.ietf.org/doc/html/rfc3779#section-2.1.1

> A prefix may be abbreviated by
> omitting the least-significant zero fields, but there should be
> enough fields to contain the indicated number of constant bits.  The
> abbreviated forms of the example IPv4 prefix is 10.5.0/23

**Comment by elrido** [comment]
Extending quote for completeness:

> For example, the set of 512 IPv4 addresses from 10.5.0.0 through 10.5.1.255 all have the same 23 most-significant bits. [...]
> A prefix may be abbreviated by omitting the least-significant zero fields, but there should be enough fields to contain the indicated number of constant bits. The abbreviated forms of the example IPv4 prefix is 10.5.0/23

I understand these rules to mean that we should not accept the even shorter 10.5/23, because we are now not able to distinguish from, for example, the range 10.5.2.0 through 10.5.3.255. 10.5/16 would be valid, as would 10.0/16, but not 10/15.

With the current changes, the library would interpret 10.5/23 as 10.0.0.5/23 and therefore expand it into the range 10.0.0.0 through 10.0.1.255, which would be wrong, while 10.5.0/23 would (happen to) get parsed correctly.

What about a logic like this in `\IPLib\Range\Subnet::fromString()` (pseudo code):
- append zeroes as necessary to IPv4 addresses, preventing the abbreviation rules for single IPs and 10.5/16 to get parsed as 10.0.0.5
- count the number of fields in the original string
- calculate the number of bits the fields can hold (8 bits per field in IPv4, 16 bits in IPv6)
- if the number of bits the fields can hold is smaller than the number of bits the subnet mask indicates, reject the range as invalid

**Comment by elrido** [comment]
Since there has been quite some changes since this MR got started, here is an overview of the current state of the changes:
- abbreviated syntax for IPv4 subnets allows for IP addresses with less than 4 dot-separated parts, as long as they contain as many or more bits as the subnet mask indicates. Now valid: `1.2.3/20`, still invalid: `1.2.3/28`
- abbreviated syntax for single IPv4 addresses allows the last dot-separated part to overflow into the missing parts, but not overwrite any given parts. Now valid: `16843009` (= `1.1.1.1`), `0.65793` (= `0.1.1.1`), `127.1` (= `127.0.0.1`), still invalid: `4294967296`, `0.0.0.257`
- abbreviated syntax still allows for parsing non decimal IPv4 parts, including overflowing ones, hence the allowed lengths got adjusted for those regexes
- the "missing dots" string manipulation only occurs once, even for subnets and only if the address is IPv4

As mentioned, the change will allow IPs that formerly got rejected. If this is not desired, I could add an optional flag for enabling this feature, similar to the `$mayIncludePort` and `$supportNonDecimalIPv4` ones. Please do let me know if you can think of any other corner cases that haven't been covered, yet.

**Comment by mlocati** [comment]
> As mentioned, the change will allow IPs that formerly got rejected. If this is not desired, I could add an optional flag for enabling this feature

I'd reuse the `$supportNonDecimalIPv4` flag.

If it's `false` we only accept the dotted-decimal format (`[0-255].[0-255].[0-255].[0-255]`), both for addresses and ranges.

If it's `true` we can accept non-dotted-decimal formats in addition to dotted-decimal.
That is, we'd also accept
- addresses represented by `[0-255].[0-255].[0-65535]`, `[0-255].[0-16777215]`, or `[0-4294967295]`
- subnet ranges represented by `[0-255].[0-255].[0-255]/[0-32]`, `[0-255].[0-255]/[0-32]`, or `[0-255]/[0-32]`
- hexadecimal and octal numbers

**Comment by mlocati** [comment]
Addenda for https://github.com/mlocati/ip-lib/pull/66#issuecomment-855264577

It seems there's no software around there that supports hexadecimal/octal subnet ranges (for example, `0x10.0.0.0/8`).

**Comment by elrido** [comment]
Thank you, this got changed. And I fixed the use of the `**` operator, which is not supported until PHP 5.6.

**Comment by elrido** [comment]
Is there anything I could still improve for this PR?

**Comment by mlocati** [comment]
@elrido Nope, everything is ok on your side. I just need to find some spare time to review everything

**Comment by mlocati** [comment]
It seems we can't parse '0xFFFFFFFF` on 32-bit systems (see https://github.com/mlocati/ip-lib/pull/66/checks?check_run_id=2996140760)

**Comment by mlocati** [comment]
Thank you for your great work, @elrido. That really helped me better understand #59 and #65, and how those issues can be fixed.

BTW we need to:
- be able to parse unsigned 32-bit integers even on 32-bit PHP instances (PHP uses 32-bit signed integers)
- avoid any backward incompatible change, so we need to add a new flag to let users specify if they accept the compact subnet notation (eg `127.0/16`) and/or the alternative IPv4 notation (eg `127.5000`).

I'm going to take care of those issues in #73
